packages/flutter_tools/lib/src/android/gradle_errors.dart

```diff
@@ -0,0 +1,299 @@
+  }) async {
+    final RunResult tasksRunResult = await processUtils.run(
+      <String>[
+        gradleUtils.getExecutable(project),
+        'app:tasks' ,
+        '--all',
+        '--console=auto',
+      ],
+      throwOnError: true,
+      workingDirectory: project.android.hostAppGradleRoot.path,
+      environment: gradleEnv,
+    );
+    // Extract build types and product flavors.
+    final Set<String> variants = <String>{};
+    for (String task in tasksRunResult.stdout.split('\n')) {
+      final Match match = _assembleTaskPattern.matchAsPrefix(task);
+      if (match != null) {
+        final String variant = match.group(1).toLowerCase();
+        if (!variant.endsWith('test')) {
+          variants.add(variant);
+        }
+      }
+    }
+    final Set<String> productFlavors = <String>{};
+    for (final String variant1 in variants) {
+      for (final String variant2 in variants) {
+        if (variant2.startsWith(variant1) && variant2 != variant1) {
+          final String buildType = variant2.substring(variant1.length);
+          if (variants.contains(buildType)) {
+            productFlavors.add(variant1);
+          }
+        }
+      }
+    }
+    printStatus(
+      '\n$warningMark  Gradle project does not define a task suitable '
+      'for the requested build.'
+    );
+    if (productFlavors.isEmpty) {
+      printStatus(
+        'The android/app/build.gradle file does not define '
+        'any custom product flavors. '
+        'You cannot use the --flavor option.'
+      );
+    } else {
+      printStatus(
+        'The android/app/build.gradle file defines product '
+        'flavors: ${productFlavors.join(', ')} '
+        'You must specify a --flavor option to select one of them.'
+      );
+    }
+    BuildEvent('flavor-undefined').send();
+    return GradleBuildStatus.EXIT;
+  },
+);
+
+/// The status of the Gradle build.
+enum GradleBuildStatus{
+  /// The tool cannot recover from the failure and should exit.
+  EXIT,
+  /// The tool can retry the exact same build.
+  RETRY,
+  /// The tool can build the plugins as AAR and retry the build.
+  RETRY_WITH_AAR_PLUGINS,
+}
+
+/// A Gradle handled error.
+///
+/// When the [test] function returns [true], [handler] is called.
+class GradleHandledError{
+  const GradleHandledError({
+    this.test,
+    this.handler,
+  });
+
+  /// The test function.
+  /// Returns [true] if the current error message should be handled.
+  final bool Function(String) test;
+
+  /// The handler function.
+  final Future<GradleBuildStatus> Function({
+    String line,
+    FlutterProject project,
+    bool usesAndroidX,
+    bool shouldBuildPluginAsAar,
+  }) handler;
+}
+
+/// Returns a simple test function that evaluates to [true] if
+/// [errorMessage] is contained in the error message.
+bool Function(String) _lineMatcher(List<String> errorMessages) {
+  return (String line) {
+    return errorMessages.any((String errorMessage) => line.contains(errorMessage));
+  };
+}
```

Comment:
generally speaking I would recommend using typedefs for these, to make it clearer.

Reply:
way better. Thanks!